contributors/design-proposals/api-machinery/admission-control-webhooks.md

```diff
@@ -211,30 +203,41 @@ When a webhook is persistently failing to allow e.g. pods to be created, then th
 When a core controller, e.g. ReplicaSet, fails to make a resources, it must send a helpful event that is visible in `kubectl describe` for the controlling resources, saying the reason create failed.
 
 
-## Conversion 
-
-The apiserver will send versioned objects to the webhook to avoid churns in the webhook code. The current registration API has an Rule.[APIVersions](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/admissionregistration/types.go#L86) field, which is used to match against the version in the endpoint url, and only matching requests will be forwarded to the webhook. The current API isn't able to specify what version the APIServer should convert the request body to before forwarding the it to the webhook. So we propose two ways to modify the API.
+## Conversion and Versioning
 
-Plan A:
+We need to make it possible to upgrade the control plane of operating clusters.
+To that end, we will make sure the communication between apiserver and webhook
+is versioned. A separate document will work out these details. Currently:
+https://docs.google.com/document/d/1BT8mZaT42jVxtC6l14YMXpUq0vZc6V5MPf_jnzDMMcg/edit
 
-We change the semantic of the APIVersions field. The field means the external versions the webhooks is able to decode. The APIServer forwards the webhook any requests that are stored at the same etcd path as the one appears in the registration API.
+It is possible that two different webhooks want to access the same
+stored-resource via different GVKs.  In this case there could be extra
+conversions.  Conversion does take some CPU.  However, we don't expect this to be
+a significant issue for beta because:
 
-Plan B:
-
-The semantic of the APIVersions field is unchanged. We add a ConvertToGroup and ConvertToVersion field to express the versions the webhooks understand. A straw-man design is [here](https://docs.google.com/document/d/1CPcRQ12dIxIE1_T8BYx41V2w5WEGP0vHAWCa7HpqsVg/edit#heading=h.8emrnnxujf1k).
-
-It's TBD if we go with Plan A or B. 
+*   Pods are by far the most common core object to webhook, based on the  [Use Cases](#use-cases-detailed-descriptions), and pods do not have multiple versions.  Most or all other resources we know we plan to Webhook also do not have multiple versions.  So extra conversion is not needed in practice.
```

Comment:
I'd like to scratch this point.

I agree that we don't really care about conversion cost of beta, but I'm not persuaded by this.  I don't think we should be optimizing for pods, but rather making an API that works in general.

Reply:
I think the point here isn't that it won't work for other cases, just that pods are likely to be a heavy target (I think you agreed at it's probably going to be at least 50% of the use) and they won't need it, therefore in the common case it's not expensive. I will change the last sentence here-- "not needed in practice" is misleading.